### PR TITLE
Simplify OrderPreservingInterner allocation strategy ~97% faster (#2677)

### LINKED ISSUE #2677: Arrow Row Format
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I think this crate has pretty good stories for operating on individual columns, either by downcasting to a concrete type, or invoking a `dyn` kernel.

The stories for multi-column operations are substantially weaker, with patchy support for common multi-column operations such as sorts, groupings, aggregations, reassembly, etc... We have some pieces such as `MutableArrayData`, `DynComparator`, but they're not especially performant, making extensive use of dynamic dispatch at the row-level, nor easy to use.

**Describe the solution you'd like**

Having a first-class row representation will not only allow us to implement more performant versions of existing kernels such as lexsort, but also provide a pretty compelling primitive to downstreams with which to implement more advanced operations such as streaming merges, joins, aggregates, etc... There is also precedent, with the C++ arrow library providing its own row format.

**Goals**

* Each row should be encoded as a single sequence of bytes
* Comparison of the byte arrays should be sufficient to establish ordering of the rows
* It should be possible to convert a selection of rows back to arrays

**Non-Goals**

* Support introspection or mutation of the row values
* Provide a stable encoding for FFI, IO, etc...
* Provide "optimal" encoding, rather a reasonable out-of-the-box baseline for common use-cases

**Describe alternatives you've considered**

We could extend the row format in DataFusion, however, this would limit its benefits to DataFusion. I think a row-oriented representation is such a fundamental primitive that it makes sense for inclusion in arrow-rs, so that it can be both used in its kernels and by downstreams that don't make use of DataFusion.

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2677

# Rationale for this change
 

Not only was the previous allocation strategy quite hard to follow, when interning sorted data it would end up comparing the value against at least `1/4` of the already interned values. This was sub-optimal.

Taking a step back the original approach of leaving spaces was to mitigate the impact of unsorted data, as the design evolved `OrderPreservingInterner` now always sorts the dictionary prior to interning. This dramatically reduces the likelihood of out-of-order inserts, and consequently diminishes the value of this mitigation.

A simpler, in-order allocation strategy is both faster and easier to understand

```
Benchmarking convert_columns 4096 string_dictionary(10, 0): Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.6s, enable flat sampling, or reduce sample count to 60.
convert_columns 4096 string_dictionary(10, 0)
                        time:   [1.1147 ms 1.1156 ms 1.1169 ms]
                        change: [-96.615% -96.612% -96.607%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

Benchmarking convert_columns 4096 string_dictionary(30, 0): Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.8s, enable flat sampling, or reduce sample count to 60.
convert_columns 4096 string_dictionary(30, 0)
                        time:   [1.1510 ms 1.1514 ms 1.1519 ms]
                        change: [-96.623% -96.621% -96.619%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Benchmarking convert_columns 4096 string_dictionary(100, 0): Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 6.1s, enable flat sampling, or reduce sample count to 60.
convert_columns 4096 string_dictionary(100, 0)
                        time:   [690.29 µs 692.36 µs 695.23 µs]
                        change: [-97.493% -97.335% -97.211%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 20 outliers among 100 measurements (20.00%)
  1 (1.00%) low mild
  19 (19.00%) high severe

convert_columns 4096 string_dictionary(100, 0.5)
                        time:   [316.77 µs 316.89 µs 317.00 µs]
                        change: [-95.710% -95.708% -95.705%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe
```

# What changes are included in this PR?

Changes the allocation strategy for OrderPreservingInterner to be simpler to understand and faster.

# Are there any user-facing changes?

No

### CODE REVIEW

File: arrow/src/row/interner.rs [hunk 1]
```diff
@@ -230,180 +226,102 @@ struct Slot {
...
-                    slot.value = Some(values_buf.insert(data))
+        let slots_len = self.slots.len() as u8;
+        // We optimise the case of inserting a value directly after those already inserted
+        // as [`OrderPreservingInterner::intern`] sorts values prior to interning them
+        match self.slots.last() {
+            Some(slot) => {
+                if &values_buf[slot.value] < data {
+                    if slots_len == 254 {
+                        out.push(255);
+                        self.next
```
Comment (alamb) [inline arrow/src/row/interner.rs:295]: Is there coverage of this branch? I think it happens when data is inserted in backwards sorted order, right? Maybe we can add at test to ensure it is covered?

**Review by alamb — APPROVED** [review]
I read the code carefully and it makes sense to me. I had one question about test coverage but otherwise 👌 

I look forward to seeing how this works

File: arrow/src/row/interner.rs (hunk 1 above)
Comment (tustvold) [reply arrow/src/row/interner.rs:295]: This branch is hit when the current bucket is "full" and it is trying to insert a value that is greater than the last value in the bucket. This therefore occurs when interning more than 255 values in order, and is therefore hit by `test_interner()` in particular

```
let mut values: Vec<_> = (0_u64..2000).collect();
test_intern_values(&values);
```
